### ISSUE TITLE
Update / create HW 1 solution

### ISSUE THREAD

**Issue by polina-tsvilodub** [CogSciPrag/Understanding-LLMs-course#21]
See provided solutions from last year for extant exercises (some might be missing).

Please note that **exercise 4** includes starting to fine-tune an LM, and therefore, is most prone to computational issues. I also vaguely remember that there were some issues with the task last year. I tried to fix the bugs that were related to the dataloader, but please critically check & stress-test this task, in particular.

**Comment by aakarsh** [comment]
Release date: 04-29-2025

**Comment by aakarsh** [comment]
Work to be reviewed by @Alishka99

**Comment by polina-tsvilodub** [comment]
Please have the HW1 tested by Apr 25th, and the solutions ready by May 13th @aakarsh @Alishka99

**Comment by Alishka99** [comment]
Exercise 3: Fine-tuning GPT-2 for QA 
1. I noticed that in the task itself it is written "Fine-tuning GPT-2 for QA" but in code we should use "EleutherAI/pythia-160m" 
2. The total score of tasks constitutes 63 points (not 50 points as written )  
3. Current tokenize helper adds inputs_ids and attention_mask to dataset with other "id","question" ,and etc data. That is why as a further step students should separately stack input_ids and attention_mask as torch dtype.

**Comment by polina-tsvilodub** [comment]
Thanks @Alishka99 ! 
1. Fixed.
2. Fixed / adjusted.
3. I added the dropping of the other columns; I think this should solve the problem with the dtype??

**Comment by aakarsh** [comment]
@polina-tsvilodub 

Still working on this but I wanted to add some preliminary comments regarding working on Exercise-4 after discussing a bit with @Alishka99 , as well, to give a heads up. Our solutions are available [here](https://drive.google.com/drive/folders/1W1CQ-8JR9szk-lQEsMC51tqDfksJgkmy?usp=drive_link). Still being finalized/fixed

1. Ran into some issues with padding longest. This makes it harder to experiment with batch size, as all inputs in a batch are expected to be of the same size.  Two alternative methods can be used to fix this, either pick a fixed padding size or ensure batch size interacts well with padding(harder). The two possible solutions are shown in the drive link.

3. Based on runs it would be nicer to use 200 steps as it is easier to debug when one does not see the standard loss curve. Would also  recommend to include a plot of what we expect the loss curve to be to give students guidance that they are missing something, I know i faced issue where a bug prevented learning but I could not discern if this was bug or just not enough steps. 

3. I believe remove columns might work, I saw the update but have not tried it, our solutions included a custom collation function.

5. It was a bit misleading calling the validation set the test set, as we don't have ground truth values for the test set. It might be more straight forward to call the the variable validation test and the curves validation curves.

**Comment by aakarsh** [comment]
I have created a draft version of the solution available [here](https://drive.google.com/file/d/1QlDqLiuk-fKP9Q-JGTvWXgvasFT0M5qZ/view?usp=drive_link). @Alishka99  , @polina-tsvilodub 

Feel free to let me know if you see some concern.

**Comment by polina-tsvilodub** [comment]
Thanks so much for your help, @aakarsh , @Alishka99! I released the homework now, incorporating your feedback. I also updated the solutions a little bit to use a predefined data collator `DataCollatorForLanguageModeling` . My working solution is in Ali's notebook that you had shared.
